### PR TITLE
[20401] Improve History QoS documentation and compatibility rules

### PR DESCRIPTION
**Pull request by JesusPoderoso**
This PR follows improves the History QoS configuration documentation and give some extra light regarding specific scenarios and compatibility with other QoS.

**Note**: This PR also updates the code / XML examples to the latest style followed in the documentation, as in continuous improvement.

Fast DDS related PR: 
- https://github.com/eProsima/Fast-DDS/pull/4375

Fixes https://github.com/eProsima/Fast-DDS/issues/4365

### CODE REVIEW

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 1]
```diff
@@ -1280,18 +1327,18 @@ Example
...
-    .. literalinclude:: ../../../../../code/DDSCodeTester.cpp
-       :language: c++
-       :dedent: 8
-       :start-after: //DDS_CHANGE_TOPIC_DATA_QOS_POLICY
-       :end-before: //!
+    .. literalinclude:: /../code/DDSCodeTester.cpp
+        :language: c++
+        :dedent: 8
+        :start-after: //DDS_CHANGE_TOPIC_DATA_QOS_POLICY
+        :end-before: //!
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:1334]: NIT space
```suggestion
       :language: c++
       :dedent: 8
       :start-after: //DDS_CHANGE_TOPIC_DATA_QOS_POLICY
       :end-before: //!
```

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 2]
```diff
@@ -1280,18 +1327,18 @@ Example
...
   .. tab:: XML
 
-    .. literalinclude:: ../../../../../code/XMLTester.xml
-       :language: xml
-       :start-after: <!-->XML_CHANGE_TOPIC_DATA_QOS_POLICY
-       :end-before: <!--><-->
+    .. literalinclude:: /../code/XMLTester.xml
+        :language: xml
+        :start-after: <!-->XML_CHANGE_TOPIC_DATA_QOS_POLICY
+        :end-before: <!--><-->
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:1341]:
```suggestion
       :language: xml
       :start-after: <!-->XML_CHANGE_TOPIC_DATA_QOS_POLICY
       :end-before: <!--><-->
```

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 3]
```diff
@@ -1351,18 +1398,18 @@ Example
...
-    .. literalinclude:: ../../../../../code/DDSCodeTester.cpp
-       :language: c++
-       :dedent: 8
-       :start-after: //DDS_CHANGE_USER_DATA_QOS_POLICY
-       :end-before: //!
+    .. literalinclude:: /../code/DDSCodeTester.cpp
+        :language: c++
+        :dedent: 8
+        :start-after: //DDS_CHANGE_USER_DATA_QOS_POLICY
+        :end-before: //!
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:1405]:
```suggestion
       :language: c++
       :dedent: 8
       :start-after: //DDS_CHANGE_USER_DATA_QOS_POLICY
       :end-before: //!
```

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 4]
```diff
@@ -1351,18 +1398,18 @@ Example
...
   .. tab:: XML
 
-    .. literalinclude:: ../../../../../code/XMLTester.xml
-       :language: xml
-       :start-after: <!-->XML_CHANGE_USER_DATA_QOS_POLICY
-       :end-before: <!--><-->
+    .. literalinclude:: /../code/XMLTester.xml
+        :language: xml
+        :start-after: <!-->XML_CHANGE_USER_DATA_QOS_POLICY
+        :end-before: <!--><-->
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:1412]:
```suggestion
       :language: xml
       :start-after: <!-->XML_CHANGE_USER_DATA_QOS_POLICY
       :end-before: <!--><-->
```

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 5]
```diff
@@ -463,24 +472,48 @@ There are two possible values (see |HistoryQosPolicyKind-api|):
...
+
+The HistoryQos must be set consistently with the :ref:`resourcelimitsqospolicy`, but also other QoS as
+:ref:`durabilityqospolicy`, so there are several cases to take into account:
+
+* The |HistoryQosPolicy::depth-api| must be consistent with the :ref:`resourcelimitsqospolicy` settings, which means
+  that the |HistoryQosPolicy::depth-api| must be equal or lower than the :ref:`resourcelimitsqospolicy`'s
+  |ResourceLimitsQosPolicy::max_samples_per_instance-api|.
+  Also, |ResourceLimitsQosPolicy::max_samples-api| must be equal or higher than the product of
+  |ResourceLimitsQosPolicy::max_samples_per_instance-api| and |ResourceLimitsQosPolicy::max_instances-api|.
+* The |HistoryQosPolicy::depth-api| cannot be zero.
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:488]: We may start here stating that the `depth` is only taken into consideration when `kind` is `KEEP_LAST`. Also, when we finally agree on the `Fast DDS's PR` how we manage having a LENGTH_UNLIMITED, we update it here.

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 6]
```diff
@@ -463,24 +472,48 @@ There are two possible values (see |HistoryQosPolicyKind-api|):
...
+  It must be equal or higher than 1, and it can be set to ``LENGTH_UNLIMITED`` to indicate that the depth is unlimited,
+  or it will be limited by the :ref:`resourcelimitsqospolicy` constraints.
+* The difference between setting the |HistoryQosPolicy::kind-api| as |KEEP_ALL_HISTORY_QOS-api| and setting it as
+  |KEEP_LAST_HISTORY_QOS-api| plus |HistoryQosPolicy::depth-api| as ``LENGTH_UNLIMITED`` is that if the
+  :ref:`durabilityqospolicy` is configured as |TRANSIENT_LOCAL_DURABILITY_QOS-api| or |TRANSIENT_DURABILITY_QOS-api|,
+  the first one will block the DataWriter when maximum :ref:`resourcelimitsqospolicy` settings reached, but the second
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:494]:
```suggestion
  the first one will block the DataWriter when maximum :ref:`resourcelimitsqospolicy` settings are reached, but the second
```

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 7]
```diff
@@ -401,18 +410,18 @@ Example
...
-    .. literalinclude:: ../../../../../code/DDSCodeTester.cpp
-       :language: c++
-       :dedent: 8
-       :start-after: //DDS_CHANGE_GROUP_DATA_QOS_POLICY
-       :end-before: //!
+    .. literalinclude:: /../code/DDSCodeTester.cpp
+        :language: c++
+        :dedent: 8
+        :start-after: //DDS_CHANGE_GROUP_DATA_QOS_POLICY
+        :end-before: //!
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:417]: NIT spacing
```suggestion
       :language: c++
       :dedent: 8
       :start-after: //DDS_CHANGE_GROUP_DATA_QOS_POLICY
       :end-before: //!
```

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 8]
```diff
@@ -401,18 +410,18 @@ Example
...
   .. tab:: XML
 
-    .. literalinclude:: ../../../../../code/XMLTester.xml
-       :language: xml
-       :start-after: <!-->XML_CHANGE_GROUP_DATA_QOS_POLICY
-       :end-before: <!--><-->
+    .. literalinclude:: /../code/XMLTester.xml
+        :language: xml
+        :start-after: <!-->XML_CHANGE_GROUP_DATA_QOS_POLICY
+        :end-before: <!--><-->
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:424]:
```suggestion
       :language: xml
       :start-after: <!-->XML_CHANGE_GROUP_DATA_QOS_POLICY
       :end-before: <!--><-->
```

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 9]
```diff
@@ -463,24 +472,48 @@ There are two possible values (see |HistoryQosPolicyKind-api|):
...
+  Also, |ResourceLimitsQosPolicy::max_samples-api| must be equal or higher than the product of
+  |ResourceLimitsQosPolicy::max_samples_per_instance-api| and |ResourceLimitsQosPolicy::max_instances-api|.
+* The |HistoryQosPolicy::depth-api| cannot be zero.
+  It must be equal or higher than 1, and it can be set to ``LENGTH_UNLIMITED`` to indicate that the depth is unlimited,
+  or it will be limited by the :ref:`resourcelimitsqospolicy` constraints.
+* The difference between setting the |HistoryQosPolicy::kind-api| as |KEEP_ALL_HISTORY_QOS-api| and setting it as
+  |KEEP_LAST_HISTORY_QOS-api| plus |HistoryQosPolicy::depth-api| as ``LENGTH_UNLIMITED`` is that if the
+  :ref:`durabilityqospolicy` is configured as |TRANSIENT_LOCAL_DURABILITY_QOS-api| or |TRANSIENT_DURABILITY_QOS-api|,
+  the first one will block the DataWriter when maximum :ref:`resourcelimitsqospolicy` settings reached, but the second
+  one will discard the oldest samples when the depth is reached.
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:495]: NIT
```suggestion
  one will discard the oldest samples when the |HistoryQosPolicy::depth-api| is reached.
```

**Review by Mario-DL — CHANGES_REQUESTED** [review]
Nice docs inclusions. Also, thanks for refactoring the C++ code tester examples in the file to the latest format.
I left some NITs and a couple of suggestions.

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst (hunk 5 above)
Comment (Mario-DL) [reply docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:488]: Finally, we should state here that `depth` cannot be less or equal than 0.

File: docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst [hunk 10]
```diff
@@ -459,28 +466,57 @@ There are two possible values (see |HistoryQosPolicyKind-api|):
...
+* Setting the |HistoryQosPolicy::kind-api| as |KEEP_ALL_HISTORY_QOS-api| uses the :ref:`resourcelimitsqospolicy`
+  limits (|ResourceLimitsQosPolicy::max_samples_per_instance-api| prior than
+  |ResourceLimitsQosPolicy::max_samples-api|).
+* In case the :ref:`reliabilityqospolicy` is set as |RELIABLE_RELIABILITY_QOS-api|, the |HistoryQosPolicy::kind-api| is
+  configured as |KEEP_ALL_HISTORY_QOS-api|, the resource limits are reached, and based on the :ref:`durabilityqospolicy`
+  configuration:
+
+  * If the |DurabilityQosPolicy::kind-api| is configured as |VOLATILE_DURABILITY_QOS-api|, the DataWriter publication
+    will discard the oldest samples.
+    Those oldest samples may belong to other :ref:`instances<dds_layer_topic_instances>`.
```
Comment (Mario-DL) [inline docs/fastdds/dds_layer/core/policy/standardQosPolicies.rst:496]: NIT
```suggestion
    Those oldest samples may belong to different :ref:`instances<dds_layer_topic_instances>`.
```

**Review by Mario-DL — APPROVED** [review]
LGTM with green CI

**Comment by EduPonz** [comment]
@Mergifyio backport 2.12.x 2.11.x 2.10.x 2.6.x

**Pull request merged**